File: Trees and Graphs/Graph.java
import java.util.HashSet;
import java.util.Deque;
import java.util.List;
import java.util.LinkedList;
import java.util.Set;

public class Graph {
  List<Node> nodes;

  // BFS
  boolean routeBetweenNodes(Node node1, Node node2) {
    Set<Node> visited = new HashSet<>();
    Deque<Node> queue = new LinkedList<>();

    // Add source node to queue
    queue.addLast(node1);

    // loop til queue is empty = we searched whole graph
    while(!queue.isEmpty()) {
      Node tmp = queue.removeFirst();
      if (tmp == node2) return true; // Found it
      visited.add(tmp);
      for (Node n : tmp.getAdjacent()) {
        if (!visited.contains(n)) {
          queue.addLast(n);
        }
      }
    }

    return false; // Not found, searched whole graph though 
  } 
}

class Node {
  private String name;
  private List<Node> adjacent;

  public Node(String name) {
    this.name = name;
  }

  public void setAdjacent(List<Node> adjacent) {
    this.adjacent = adjacent;
  }

  public List<Node> getAdjacent() {
    return this.adjacent;
  }
}

File: Data structures/Swift/DataStructure/DataStructure/Graph/Graph.swift
//
//  Graph.swift
//  DataStructure
//
//  Created by Đinh Văn Nhật on 2019/11/16.
//  Copyright © 2019 Đinh Văn Nhật. All rights reserved.
//


public struct Graph<T> where T: Hashable {

    public var V: Int
    public var E: Int

    public init() {
        self.V = 0
        self.E = 0
    }

    public mutating func addUndirectedEdge(_ data1: T, _ data2: T) {
        let node1 = getNode(data1) ?? Node(data: data1)
        let node2 = getNode(data2) ?? Node(data: data2)
        node1.adj.append(node2)
        node2.adj.append(node1)
        nodes.updateValue(node1, forKey: data1)
        nodes.updateValue(node2, forKey: data2)
        self.E += 1
    }

    public func bfs(_ src: T, _ des: T) -> Bool {
        // get node src, get node des
        guard let source = nodes[src], let destination = nodes[des] else {
            return false
        }

        var visited = Set<Node<T>>()

        // create a queue for BFS purpose
        var q = LinkedList<Node<T>>()

        // enqueue src node
        q.append(source)

        // while queue is not empty try to
        while !q.isEmpty() {
            // dequeue get first element
            let first = try! q.removeFirst()
            // for each adjacency node of first
            for adj in first.adj {
                // if this is the node we are looking for => return true, since we found it
                if adj == destination {
                    return true
                }
                // if element is not traversed - color = 0
                if visited.contains(adj) {
                    // change adjacency node color to 1
                    visited.insert(adj)
                    // enqueue
                    q.append(adj)
                }
            }
        }
        // return false since we searched hole graph
        return false
    }

    func getNode(_ data: T) -> Node<T>? {
        return nodes[data]
    }

    // MARK: - private variables

    private var nodes = [T: Node<T>]()
}

extension Graph: CustomStringConvertible {
    public var description: String {
        var str = "V = \(self.V), E = \(self.E)\n"
        for node in nodes.values {
            str += "\(node.data) -> "
            str += "\(node.adj)\n"
        }
        return str
    }
}
